### PR TITLE
feat: set user agent in webdav requests

### PR DESCRIPTION
**Pull request by lukevmorris**
I have a WebDAV server that I use as a backend remote cache for multiple build tools. Many of these build tools (e.g. `bazel`, `gradle`) differentiate themselves in their user agent header, whereas `sccache` does not. This PR configures the WebdavCache service with an http client preconfigured with a user agent header like so: `sccache/0.8.2`. These changes are nearly identical to those in https://github.com/mozilla/sccache/pull/2137. Happy to make whatever changes you feel may be necessary. Thanks!

### CODE REVIEW

File: src/cache/webdav.rs [hunk 1]
```diff
@@ -32,11 +34,19 @@ impl WebdavCache {
...
 
         let op = Operator::new(builder)?
             .layer(LoggingLayer::default())
             .finish();
         Ok(op)
     }
 }
+
+/// Set the user agent (helps with monitoring on the server side)
+fn set_user_agent() -> HttpClient {
```
Comment (Xuanwo) [inline src/cache/webdav.rs:48]: I recall that we already have similar code that builds an HTTP client. Would you like to check the codebase again?

Comment (lukevmorris) [reply src/cache/webdav.rs]: You're right! I've removed the copied code and referenced `set_user_agent` from `src/cache/s3.rs`.

**Review by Xuanwo — COMMENTED** [review]
Others look good to me. Thank you @lukevmorris for working for this.

File: src/cache/s3.rs [hunk 2]
```diff
@@ -68,7 +68,7 @@ impl S3Cache {
 }
 
 /// Set the user agent (helps with monitoring on the server side)
-fn set_user_agent() -> HttpClient {
+pub fn set_user_agent() -> HttpClient {
```
[the remark this answers is not in the document]
Comment (lukevmorris) [reply src/cache/s3.rs]: Done -- apologies if there are cleaner ways to do this, I'm not very familiar with Rust.

**Review by Xuanwo — APPROVED** [review]
Thank you @lukevmorris for working on this!

**Comment by sylvestre** [comment]
Could you please document this ?

https://github.com/mozilla/sccache/?tab=readme-ov-file
just a small section like
"sccache queries are sent with user agent XXX"
etc

**Comment by sylvestre** [comment]
thanks.
Sorry for the dumb question. I am wondering why it should not be done for all backend? Why limiting only s3 & webdav ?

(your patch is conflicting)

**Comment by lukevmorris** [comment]
@sylvestre not a dumb question at all. I had considered that but wasn't sure what scope of change would be appreciated. I can try and make that change if you'd like?

**Comment by sylvestre** [comment]
yeah, i think it makes sense :)

**Comment by lukevmorris** [comment]
Ok I ended up moving `set_user_agent()` into its own file that's included for only specific features, to try and clean up all my `cfg` blocks. Every opendal service backed by an http client is configured to use the same user agent. Redis and Memcached don't accept an http_client, I assume because they have their own specific communication protocol, so I've omitted those from this changeset. I've also added a line of documentation to readmes for each of the affected services.

Let me know what you think!

**Comment by lukevmorris** [comment]
*facepalm* I didn't check in the new file. Hopefully that addresses the test failures.

File: docs/Azure.md [hunk 3]
```diff
@@ -1,9 +1,11 @@
...
 
-To use Azure Blob Storage, you'll need your Azure connection string and an _existing_ Blob Storage container name.  Set the `SCCACHE_AZURE_CONNECTION_STRING`
-environment variable to your connection string, and `SCCACHE_AZURE_BLOB_CONTAINER` to the name of the container to use.  Note that sccache will not create
+To use Azure Blob Storage, you'll need your Azure connection string and an _existing_ Blob Storage container name. Set the `SCCACHE_AZURE_CONNECTION_STRING`
+environment variable to your connection string, and `SCCACHE_AZURE_BLOB_CONTAINER` to the name of the container to use. Note that sccache will not create
 the container for you - you'll need to do that yourself.
 
 You can also define a prefix that will be prepended to the keys of all cache objects created and read within the container, effectively creating a scope. To do that use the `SCCACHE_AZURE_KEY_PREFIX` environment variable. This can be useful when sharing a bucket with another application.
 
+Requests sent to Azure Blob Storage will have a user agent header indicating the current sccache version, e.g. `sccache/0.8.2`.
```
Comment (sylvestre) [inline docs/Azure.md:9]: i don't think we should duplicate it
just declare it once :)

Comment (lukevmorris) [reply docs/Azure.md]: Ok sorry! I centralized it in one section at the bottom of the README. What do you think?

**Pull request merged**